### DISCUSSION TITLE
mock-array placement is very different from row to row

### DISCUSSION THREAD

**Discussion by oharboe** [The-OpenROAD-Project/OpenROAD#3745] [General]
I'm curious as to why the placement gives so different results for a situation that should be identical.

The rows in mock-array are identical and there are io_lsbs_* on the right of mock-array that are fed flip flops.

A snapshot from `make gui_place`:

![image](https://github.com/The-OpenROAD-Project/OpenROAD/assets/2798822/91c718ad-7392-4af6-8ed4-ce2184e6ef80)

If I zoom in on the upper right row, I see that the io_lsbs_*, fed by a flipflop that is placed right next to the outputs are somewhat spread out:

![image](https://github.com/The-OpenROAD-Project/OpenROAD/assets/2798822/25d2393c-d316-4ca0-a50c-9bdeef629066)

If I look at the row below and to the right, I see that the io_lsbs_* are clustered together, which looks neat.

![image](https://github.com/The-OpenROAD-Project/OpenROAD/assets/2798822/f5530a21-1135-4ad3-9f60-3304f2538d2f)

At the bottom row to the right, the io_lsbs_* are mixed into the other pins:

![image](https://github.com/The-OpenROAD-Project/OpenROAD/assets/2798822/5b4d25a7-567f-480f-bf58-736cb65bb0e3)

Side-note: make foo_issue doesn't work out of the box on mock-array because the size of the tar.gz file is ca. 100mBytes. This complicates filing bug reports and feature requests.

Side-note: the io_lsbs_* are driven by a flip flop, then go to an output cell and are finally buffered before they go to the pin of the macro. Silly question: what does an output cell do and why does it need to be buffered? Can't the flip-flop drive the output pin directly?

**Comment by maliberty** [comment]
Global placement isn't a process that will produce identical results in all areas, even if they have symmetry.  We could look at specific areas if they seem to be suboptimal.

You can put the issue in google drive and put a link in the issue.

We buffer the inputs and outputs by default but it isn't strictly mandatory.  FF often don't have a large amount of drive strength for driving output signals.  However with no output loading in the sdc it probably doesn't matter but is a bit unrealistic.

Btw, I still find there is too much hold buffering due to the unrealistically io constraints.

**Comment by oharboe** [comment]
> Global placement isn't a process that will produce identical results in all areas, even if they have symmetry. We could look at specific areas if they seem to be suboptimal.

I was just curious how much difference is considered pathological. If, looking at the pictures, this variation is expected, then I don't think there is anything more to investigate here.

`make global_place_issue` generated a 120mByte .tar.gz file, but zipping just the 3_place.odb file gave me 2 mByte .zip file: [place.zip](https://github.com/The-OpenROAD-Project/OpenROAD/files/12225008/place.zip)

To generate it locally, run `make DESIGN_CONFIG=designs/asap7/mock-array/config.mk place`

If there is something interesting here, then this discussion can be converted to a feature request with the click of a button...

> You can put the issue in google drive and put a link in the issue.
> 
> We buffer the inputs and outputs by default but it isn't strictly mandatory. FF often don't have a large amount of drive strength for driving output signals. However with no output loading in the sdc it probably doesn't matter but is a bit unrealistic.

I imagine that a flip flop with a lot of drive strength would be bigger and slower... seems sensible to create a small, fast and weak flip flop and leave buffering to buffers and inverters...

> 
> Btw, I still find there is too much hold buffering due to the unrealistically io constraints.

Yes... I've got changes locally removes that, but I need to understand the clock tree generation better first. That's separate concern to placement though.

**Comment by maliberty** [comment]
If I look at the current mock array the areas around the IOs are so filled with hold cells that the variation isn't as interesting as you show above as they are packed with cell, eg
![image](https://github.com/The-OpenROAD-Project/OpenROAD/assets/761514/85c19478-10dc-4e52-98e0-79af467df0f0)

Also the IOs are not placed symmetrically with respect to the macros which will influence global placement.

**Reply by oharboe** [reply]
Makes sense w.r.t. the hold cells. Could you rearticulate "Also the IOs are not placed symmetrically with respect to the macros which will influence global placement.", I'm not quite sure what you mean.

You mean that the pins should be placed to get horizontal/vertical wires?

**Reply by maliberty** [reply]
I mean the distance from the top pin to the top of the macro varies, eg
![image](https://github.com/The-OpenROAD-Project/OpenROAD/assets/761514/4177a78b-c5ce-4615-a9e3-83bbc1fec83e)

So in some cases you see 
![image](https://github.com/The-OpenROAD-Project/OpenROAD/assets/761514/4240aedd-61e2-4778-8cd8-087b90bd1e6d)

while in others
![image](https://github.com/The-OpenROAD-Project/OpenROAD/assets/761514/242199ce-2395-447f-a4a7-32430472b2b3)

**Comment by oharboe** [comment]
If this behavior exists when the hold cells are gone, then it's worth a second look.
